### PR TITLE
Join empty lines with only one space in `join_selections`

### PR DESCRIPTION
**Pull request by TornaxO7**
fixes the issue that lines with only spaces are getting joined as well

fixes https://github.com/helix-editor/helix/issues/8977

### CODE REVIEW

File: helix-term/src/commands.rs [hunk 1]
```diff
@@ -4390,28 +4390,39 @@ fn format_selections(cx: &mut Context) {
...
+        let text = doc.text();
 
-        changes.reserve(lines.len());
+        let selected_lines = {
+            let (start, mut end) = selection.line_range(slice);
+            if start == end {
+                end = (end + 1).min(text.len_lines() - 1);
+            }
+            start..end
+        };
```
Comment (the-mikedavis) [inline helix-term/src/commands.rs:4407]: Could you reset the unnecessary changes like variable renames and moving the `slice` and `text` bindings? Inlining `Tendril::from(" ")` is ok since that will avoid some allocations but the rest is just cosmetic

File: helix-term/src/commands.rs [hunk 2]
```diff
@@ -4390,28 +4390,39 @@ fn format_selections(cx: &mut Context) {
...
+        for selected_line in selected_lines {
+            let start = line_end_char_index(&slice, selected_line);
+            let mut end = text.line_to_char(selected_line + 1);
             end = skip_while(slice, end, |ch| matches!(ch, ' ' | '\t')).unwrap_or(end);
 
             // need to skip from start, not end
-            let change = (start, end, Some(fragment.clone()));
+            let change = {
+                let separator = {
+                    let line_contains_only_space = text.char(end) == '\n';
```
Comment (the-mikedavis) [inline helix-term/src/commands.rs:4416]: This check won't work for non-LF line ending documents. For example if you try the reproduction case from the issue with `:line-ending crlf` and then `JJ` it will still have the empty space.

**Comment by the-mikedavis** [comment]
This would be a good opportunity to add a test case for `J` to the integration tests in `helix-term/tests/test/commands.rs`

File: helix-term/src/commands.rs (hunk 1 above)
Comment (TornaxO7) [reply helix-term/src/commands.rs]: updated. I'd like to have this block at least:
```rust
        let lines = {
            let (start, mut end) = selection.line_range(slice);
            if start == end {
                end = (end + 1).min(text.len_lines() - 1);
            }
            start..end
        };
```
makes it more readable, in my opinion, since we are seeing that we just need `stard` and `end` for `lines` but I can remove it if you insist.

File: helix-term/src/commands.rs (hunk 2 above)
Comment (TornaxO7) [reply helix-term/src/commands.rs]: updated

**Comment by TornaxO7** [comment]
Damn, hm... may I ask if there's a possible way to start the tests for the other platforms?

**Comment by TornaxO7** [comment]
ok, just found [act](https://github.com/nektos/act) and I'll try with it.

File: helix-term/src/commands.rs (hunk 1 above)
Comment (pascalkuthe) [reply helix-term/src/commands.rs]: I think this kind of extra indention ist actually very unreadable and clutterwd. Please reverse the Extra nestong jere and also remove the unnecssary nestong in the cide you added.

Comment (the-mikedavis) [reply helix-term/src/commands.rs]: Yeah in general I'm not a big fan of the `let var = { .. }` syntax unless we're avoiding some really nasty scoping issues. Readability / cosmetic changes like this are always debatable and they inflate the diff (make review harder and might introduce unnecessary conflicts with other PRs) so we try to avoid them

Comment (TornaxO7) [reply helix-term/src/commands.rs]: reverted

**Comment by TornaxO7** [comment]
@the-mikedavis ok, could you please restart the workflow please? Hopefully it will work this time.

File: helix-term/src/commands.rs [hunk 3]
```diff
@@ -4411,7 +4410,19 @@ fn join_selections_impl(cx: &mut Context, select_space: bool) {
...
+                    if line_contains_only_space {
+                        None
+                    } else {
+                        Some(Tendril::from(" "))
+                    }
+                };
+
+                (start, end, separator)
+            };
             changes.push(change);
```
Comment (the-mikedavis) [inline helix-term/src/commands.rs:4426]: Just some style nits:

* I'd like to avoid the extra nesting here too
* Turning an `if` condition into a var with a good name is a nice tool for long `if` conditions but IMO it's always clearer to just have a comment explaining what the `if` is branching on:

```rust
let separator = if end == line_end_char_index(&slice, line + 1) {
    // The line only contains whitespace so skip adding a space.
    None
} else {
    Some(Tendril::from(" "))
};
// need to skip from start, not end
changes.push((start, end, separator));
```

**Review by the-mikedavis — COMMENTED** [review]
This is looking good :+1:

Just some small style nits

**Review by the-mikedavis — APPROVED** [review]
Thanks for fixing this!

**Review by pascalkuthe — APPROVED** [review]

**Comment by cezhunter** [comment]
is there a way to join lines *without* including spaces? Say you have something like
```
func
    .test()
    .evaluate()
    .whatever()
```
I want to join the lines into one but not have it be `func .test() .evaluate() .whatever()`

**Comment by TornaxO7** [comment]
Hm... we could introduce a special case: If the current line starts with a punctuation, the whitespace of the previous line should be omitted. The other question is: Is there a case where the whitespace shouldn't be removed if the current line starts with a punctuation?

**Comment by the-mikedavis** [comment]
For that case you can use `A-J` (<kbd>Alt</kbd>+<kbd>Shift</kbd>+<kbd>j</kbd>) which joins lines but selects the separator. Then you can use `d` to delete the spaces.

**Comment by cezhunter** [comment]
Yup - actually figured it out shortly after I commented (`Alt-Shift-J` and `d`). That is something I like about Helix - if what you expect doesn't work try it with `shift` or `alt` :) same thing with paste (`shift` lets you paste without the newline)
